src/core/basic_client.rs

```diff
@@ -180,25 +179,73 @@ use zeroize::Zeroizing;
     ///
-    /// Before you can use this client for cryptographic operations, you first need to call
-    /// [`set_implicit_provider`](#method.set_implicit_provider). In order to get a list of
-    /// supported providers, call the [`list_providers`](#method.list_providers) method.
+    /// The client will be initialised with default values obtained from the service for the
+    /// implicit provider and for application identity.
+    ///
+    /// * `app_name` is the application name to be used if direct authentication is the default
+    /// authentication choice
     ///
     /// This client will use the default configuration. That includes using a Protobuf converter
     /// for message bodies and a Unix Domain Socket IPC handler.
-    pub fn new(auth_data: AuthenticationData) -> Self {
-        BasicClient {
+    ///
+    /// # Errors
+    ///
+    /// The errors that can be expected are all the ones coming from
+    /// [`set_default_auth`](#method.set_default_auth) and
+    /// [`set_default_provider`](#method.set_default_provider)
+    pub fn new(app_name: Option<String>) -> Result<Self> {
+        let mut client = BasicClient {
             op_client: Default::default(),
-            auth_data,
-            implicit_provider: None,
+            auth_data: AuthenticationData::None,
+            implicit_provider: ProviderID::Core,
+        };
+        client.set_default_provider()?;
+        client.set_default_auth(app_name)?;
+        Ok(client)
+    }
+
+    /// Query the service for the list of authenticators provided and use the first one as default
+    ///
+    /// * `app_name` is to be used if direct authentication is the default choice
+    ///
+    /// # Errors
+    ///
+    /// If no authenticator is reported by the service, a `NoAuthenticator` error kind is returned.
+    ///
+    /// If the default authenticator is `DirectAuthenticator` and `app_name` was set to `None`,
+    /// an error of type `MissingParam` is returned.
+    ///
+    /// If none of the authenticators returned by the service is supported, `NoAuthenticator` is
+    /// returned.
+    pub fn set_default_auth(&mut self, app_name: Option<String>) -> Result<()> {
+        let authenticators = self.list_authenticators()?;
+        if authenticators.is_empty() {
+            return Err(Error::Client(ClientErrorKind::NoAuthenticator));
         }
+        for authenticator in authenticators {
+            match authenticator.id {
+                AuthType::NoAuth => self.auth_data = AuthenticationData::None,
+                AuthType::Direct => {
+                    self.auth_data = AuthenticationData::AppIdentity(Secret::new(
+                        app_name.ok_or(Error::Client(ClientErrorKind::MissingParam))?,
+                    ))
+                }
+                AuthType::UnixPeerCredentials => {
+                    self.auth_data = AuthenticationData::UnixPeerCredentials
+                }
+                _ => continue,
```

Comment:
We could have a specific log message without having to have one match branch per type - just capturing and printing the `AuthType` that was not supported. Apart from that I don't think this failing is of much use, since there's no guarantee that when we add the new variant in the interface is also when we'll implement the functionality for it.

Reply:
I've added a log, but not a separate arm for each type